ext/dom/php_dom.c

```diff
@@ -1519,7 +1520,14 @@ static zval *dom_nodelist_read_dimension(zend_object *object, zval *offset, int
 		return NULL;
 	}
 
-	ZVAL_LONG(&offset_copy, zval_get_long(offset));
+	zend_long lval;
+	if (Z_TYPE_P(offset) == IS_LONG) {
+		ZVAL_LONG(&offset_copy, Z_LVAL_P(offset));
+	} else if (Z_TYPE_P(offset) == IS_STRING && is_numeric_string(Z_STRVAL_P(offset), Z_STRLEN_P(offset), &lval, NULL, false) == IS_LONG) {
+		ZVAL_LONG(&offset_copy, lval);
+	} else {
+		return NULL;
```

Comment:
You can't just return NULL here, because in that case `read_dimension()` expects to get an exception.
i.e. the following code (from the bug report) will crash PHP with an assertion failure in debug mode:

```php
<?php

$doc = new DOMDocument;
$doc->loadHTML('<a href="https://example.com">hi!</a>');
$a = (new DOMXPath($doc))->query('//a')[0];
$a->attributes['href']->nodeValue = 'https://example.net';
print $doc->saveHTML();
```